app/Domain/Timesheets/Services/Timesheets.php

```diff
@@ -231,13 +301,119 @@ public function getAll(Carbon $dateFrom, Carbon $dateTo, int $projectId = -1, st
      *
      * @return array
      */
-    public function getWeeklyTimesheets(int $projectId, Carbon $fromDate, int $userId = 0): array
+    public function getWeeklyTimesheets(int $projectId, CarbonImmutable $fromDate, int $userId = 0): array
     {
-        return $this->timesheetsRepo->getWeeklyTimesheets(
+
+        //Get timesheet entries and group by day
+
+        $allTimesheets = $this->timesheetsRepo->getWeeklyTimesheets(
             projectId: $projectId,
             fromDate: $fromDate,
             userId: $userId
         );
+
+        //Timesheets are grouped by ticketId + type
+        $timesheetGroups = [];
+        foreach ($allTimesheets as $timesheet) {
+
+            $currentWorkDate = (new DateTimeHelper())->parseDbDateTime($timesheet['workDate']);
+
+            //Various Entries can be in different timezones and thus would not be caught by upsert or grouping by default
+            //Creating new rows for each timezone adjustment
+            $timezonedTime = $currentWorkDate->format("H:m:s");
+
+            $groupKey = $timesheet["ticketId"] . "-" . $timesheet["kind"] . "-". $timezonedTime;
+
+            if (!isset($timesheetGroups[$groupKey])) {
+                //Build an array of 7 days for the weekly timesheet. Include start date of the current users timezone in
+                //UTC. That way we can compare the dates coming from the db
+                $timesheetGroups[$groupKey] = array(
+                    "kind" =>  $timesheet["kind"],
+                    "clientName" => $timesheet["clientName"],
+                    "name" => $timesheet["name"],
+                    "headline" => $timesheet["headline"],
+                    "ticketId" => $timesheet["ticketId"],
+                    "day1" => array(
+                        "start" =>  $fromDate,
+                        "end" => $fromDate->addHours(23)->addMinutes(59),
```

Comment:
We not use `endOfDay()` here?

Reply:
Great question: In this case we are not actually looking for the end of the day but 24hours in the future. Since this is the UTC date the date time could be starting on Sunday 18:00. So then day1 should be Sunday 18:00 to Monday 17:59 UTC. 

The reason for this range is small timezone shifts will break the weekly timesheets. For example daylight savings time or just traveling to a different timezone might add a new UTC date to the db which does not fully align with the datetime in the previous zone:

Example:
Monday: adding 4 hours while in New York
Tuesday: Traveled to San Francisco (3 hours back) and forgot to add another 2 hours to Monday. So I add a new entry. 

Because dates are stored as start of day 00:00:00 in the users timezone we will have 2 rows in the db one for Monday, 4 hours at UTC-New York and one for 2 Hours at UTC-San Fran. 

The grouping  was not taking that scenario into consideration. Now it does and just uses the users current timezone start of day transformed to UTC and then adds 24 hours. 

* I hate Timezone Management *